File: apps/spell-circle-canvas/src/common/compose/reference/examples/textStroke_verb.py
"""textStroke — the letterforms thickened by a pass beneath their fill,
which is what keeps a caption legible over a busy ground.

The Python twin of textStroke_verb.cpp. A reference example: it is
rendered with ``sigil render`` and belongs to no sketch registry.
"""

from sigil import compose
from sigil.sketch import SketchContext, sketch

GROUND = "#14181d"
BUSY = "#6a8f7f"
INK = "#f4f7f9"
OUTLINE = "#121a1e"
ASH = "#8ea0ad"


def cell(caption: str, label: compose.Element) -> compose.Element:
    """The same word over the same ground, once plain and once
    engraved."""
    return (
        compose.box(
            compose.box(label)
            .width(compose.pct(100))
            .height(120)
            .borderRadius(10)
            .fill(BUSY)
            .justifyContent("center")
            .alignItems("center"),
            compose.text(caption, size=12, color=ASH),
        )
        .column()
        .gap(10)
        .flexBasis(0)
        .flexGrow(1)
        .alignItems("center")
    )


@sketch(size=(600, 250), background=GROUND, capture_at=0)
class TextStrokeVerb:
    def setup(self, ctx: SketchContext) -> None:
        ctx.render(self.describe())

    def describe(self) -> compose.Element:
        return (
            compose.box(
                cell(
                    "the letterforms alone",
                    compose.text("LEGIBLE", size=34, color=INK).fontTrack(1),
                ),
                cell(
                    "textStroke(4, outline)",
                    compose.text("LEGIBLE", size=34, color=INK)
                    .fontTrack(1)
                    .textStroke(4, OUTLINE),
                ),
            )
            .row()
            .gap(18)
            .padding(24)
        )

File: apps/spell-circle-canvas/src/common/compose/reference/examples/textStroke_verb.cpp
/** @file
 * textStroke — the letterforms thickened by a pass beneath their fill,
 * which is what keeps a caption legible over a busy ground.
 *
 * A reference example. It stands outside the sketch registry, so it is
 * photographed with `--frame` and never enters the plate sweep.
 */

#include <sigilcompose/core/Core.h>
#include <sigilsketch/canvas/Sketch.h>

namespace sketch = sigil::sketch;

using namespace sigil::compose;

namespace {

constexpr SkSize kCanvas = {600, 250};
constexpr SkColor4f kGround = hexColor(0x14181d);
constexpr SkColor4f kBusy = hexColor(0x6a8f7f);
constexpr SkColor4f kInk = hexColor(0xf4f7f9);
constexpr SkColor4f kOutline = hexColor(0x121a1e);
constexpr SkColor4f kAsh = hexColor(0x8ea0ad);

/** The same word over the same ground, once plain and once engraved. */
Element cell(const char* caption, Element label) {
  return box()
      .column()
      .gap(10)
      .basis(0)
      .grow(1)
      .alignItems(Align::Center)
      .children({box()
                     .width(pct(100))
                     .height(120)
                     .borderRadius({10})
                     .fill(kBusy)
                     .justify(Justify::Center)
                     .alignItems(Align::Center)
                     .children({std::move(label)}),
                 text(caption).font({.size = 12, .color = kAsh})});
}

}  // namespace

struct TextStrokeVerb {
  void setup(sketch::SketchContext& ctx) {
    ctx.canvas({.size = kCanvas, .background = kGround, .captureSeconds = 0});
    ctx.composer.render(describe());
  }

  Element describe() const {
    return box().row().gap(18).padding(24).children({
        cell("the letterforms alone",
             text("LEGIBLE").font({.size = 34, .color = kInk, .track = 1})),
        cell("textStroke(4, outline)",
             text("LEGIBLE")
                 .font({.size = 34, .color = kInk, .track = 1})
                 .textStroke(4, Fill::color(kOutline))),
    });
  }
};

SIGIL_SKETCH(TextStrokeVerb, "Reference · Compose",
             "a caption over a busy ground, with and without the stroke "
             "under its fill")
